Adding docs for elastic-charts

cchaos opened the pull request:
This PR will add docs pages for 4 main topics surrounding elastic-charts (and charts in general).

The sections will show the **ideal** version of the specified chart type. They talk more about how to represent your data correctly than how elastic-charts works. That's what **their** docs are for. However, it does also  provide code to produce the resulting chart so it's easy for consumers to copy/paste these **ideal** charts.

### 1. Using elastic-charts and EUI together

Main topic is how to use the EUI produced themes that work with elastic-charts and then a broader topic on best uses for color when creating charts.

### 2. Time series charts

This is probably the most used chart type in Kibana and will then probably be the most consumed version of elastic-charts.

### 3. Categorical charts

This is a type of chart that usually gets misrepresented because consumers don't adjust their data properly.


### 4. Sizing/Sparklines 

There isn't always room to show all the details of a chart depending on the size of the chart itself. The sizing section showcases some examples.

Most of the issues that stem around the creation of sparklines is forgetting to remove interactions and extraneous markings. 

### Checklist

- [x] Checked in **dark mode**
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- ~[ ] Props have proper **autodocs**~
- [x] Added **documentation** examples
- ~[ ] Added or updated **jest tests**~
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

markov00 commented:
Thanks @cchaos for this, it looks awesome.

I have some small comments on the code, but I think I can go through them and fix them directly, like some code strings instead of the `Position` object, some snippet code has the `xScaleType` on the `Axis`.

I think we should also include an example to hide the markers on the line if the line contains many points (100 or more).

My last remark: the font size seems very small for all the labels on the chart, don't you think? Compared to the basic font size (14px)?

cchaos commented:
Thanks @markov00. Yeah I went through a pass to clean up those extra objects but must have missed some.

>I think we should also include an example to hide the markers on the line if the line contains many points (100 or more).

Do you think this should just be a part of elastic-charts? Something it does automatically so that the marker only exists on hover?

> the font size seems very small for all the labels on the chart, don't you think? 

Agreed. I originally set it that small to sort of _get them out of the way_ because charts should just be a quick glance of noticing trends and differences and then the user should rely on the tooltips for actual values. 

The current problem right now is that elastic-charts doesn't allow for good spacing between the tick labels causing a jam up of text.

<img src="https://d.pr/free/i/zxiiqK+" />

So reducing the font size helped to keep the user from getting distracted by these values. I do realize, however, that the font size is too small for categorical charts where reading the category names are important for that glance. And there still needs to be better vertical spacing.

<img src="https://d.pr/free/i/QNsTRs+" width="50%" />

> Compared to the basic font size (14px)?

The base font size for charts has never been as large as 14px. 

<img src="https://d.pr/free/i/19sKsF+"  />

But I'm thinking a good compromise is increasing both ticks and labels by just a couple pixels.

<img src="https://d.pr/free/i/fWhEuS+" />

markov00 commented:
@cchaos
> Do you think this should just be a part of elastic-charts? Something it does automatically so that the marker only exists on hover?

Do we want to hardcode that max number of markers on a specific size or do we want to add that to the theme itself? I think the last should provide all a better configurability of the chart

For the axis labels: we have a ticket to fix too many labels specifying the number of ticks for each axis.
You can also use the `padding` style of the `tickLabelStyle` to allow more space between labels

cchaos commented:
This PR is now ready for review 🎉

cchaos commented:
@markov00 

> Do we want to hardcode that max number of markers on a specific size or do we want to add that to the theme itself? I think the last should provide all a better configurability of the chart

I was thinking that since the chart is already doing all that math and plotting of points that it should be able to determine a minimum distance between points and if that minimum is not met, it would turn off the points. Of course this could be a boolean style prop or setting that just turns the logic on or off but is on by default.

I'm not sure any other good way to do it through the theme (at least with the current configurations that are available now).

cchaos commented:
@snide 

> Possibly some more instruction on the palette services (chroma) being used. Just something that says... to build charts that look like this. Import this theme, use this service, and then use this chart configuration. Some of that is lost in the doc examples at the moment.

This is currently not the final way I'd like to be able to support more color schemes/palettes. We can either provide a better service with our own functions (pulling the chroma one out of docs) on top of our named palettes. [Or Elastic-charts needs to create a color service for creating palettes](https://github.com/elastic/elastic-charts/issues/293).

Either way, there's no good way to describe the way I'm creating these extra palettes at the moment. My biggest goal right now is to **get the base EUI chart themes into Kibana** so that all the charts can start to look a bit better. I don't think many are using custom color schemes, or if they are they're already providing these hex values directly.

snide commented:
Yes. I agree. It should be handled on the ECharts side.

thompsongl commented:
The panels and configuration toggles with built-in best practices is a _very cool_ and _very helpful_ way to do this.
These look to be some of the best examples we have

snide commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -0,0 +1,240 @@
+      <Fragment>
+        <EuiTitle size="xxs">
+          <h3>
+            Number of GitHub issues per visualization type
```
Looks like the legend has a small bug. It's reporting the hover values rather than an aggregation?

![](http://snid.es/4ff6c942a86c/Screen%252520Recording%2525202019-08-19%252520at%25252011.13%252520PM.gif)

markov00 replied:
There is no "aggregation" value on the legend to be applied. 
We for sure have to remove the value when not on hover (it's used on tsvb to show the last bucket but for  time series, for ordinal values is not useful).

thompsongl commented on src-docs/src/views/elastic_charts/theming_example.js:
```diff
@@ -0,0 +1,122 @@
+            pass the correct one to the Settings.theme property.
+          </p>
+          <EuiCodeBlock language="javascript" isCopyable>
+            {`import { EUI_DARK_THEME, EUI_LIGHT_THEME } from \'@elastic/eui/src/themes/charts/themes\';
```
Yes, these should come from `dist`

thompsongl replied:
But this will require some webpack work. I can look into it

thompsongl commented:
PR adding the chart themes to `dist`: cchaos/eui#23

gchaps commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -0,0 +1,240 @@
+
+          <EuiFlexItem>
+            <ChartTypeCard
+              type="Only bar charts are recommended but categorical"
```
Might need some tweaking:

Although we recommend only bar charts, you can use any categorical chart for X/Y series data.

cchaos replied:
This is a tough one because it has to end in `charts can be displayed as any x/y series type.` the way that I'm re-using that particular card. Maybe, `Although we recommend only bar charts, categorical charts can be displayed as any x/y series type.`

gchaps commented on src-docs/src/views/elastic_charts/theming_example.js:
```diff
@@ -0,0 +1,122 @@
+      demo: <Theming />,
+    },
+    {
+      title: 'Coloring charts correctly',
```
Remove correctly. It's implied that guidelines are all about doing things correctly.

ryankeairns commented:
Just stopping by to say that I pulled this down and it is awesome. Really great work @cchaos !

cchaos commented on src-docs/src/views/elastic_charts/category_chart.js:
```diff
@@ -0,0 +1,240 @@
+              description="Tick marks should be spaced out properly and number values should be formatted. For example, if the number is in the thousands, remove a few numerals and add the `k` symbol.">
+              <EuiCode>1000 ⇢ 1k</EuiCode> &nbsp; <EuiCode>20000 ⇢ 20k</EuiCode>
+              <EuiSpacer size="s" />
+              <EuiSwitch
```
No... but maybe I should convert it

cchaos replied:
Changed this so that it actually is in the thousands.

cchaos commented on src/themes/charts/themes.js:
```diff
@@ -0,0 +1,125 @@
+import { palettes } from '../../services/color/eui_palettes';
+import { DEFAULT_VISUALIZATION_COLOR } from '../../services/color/visualization_colors';
+
+import lightColors from '!!sass-vars-to-js-loader!../../global_styling/variables/_colors.scss';
```
@thompsongl With the webpack alt that you provided, will it export the actual compiled JSON or does it just copy this file to the dist?

thompsongl replied:
The webpack build now compiles this file to `dist` as a commonjs. So it's not JSON, but the themes will come in a JS objects

cchaos replied:
Sweet. I just rand a build too to make sure there were no errors and the build works. Thanks!!

cchaos replied:
@thompsongl So I'm testing out the build in Kibana to make sure I can get to these theme files and it works just fine, but I'm just getting a missing declaration file error. Is this something we fix on our end or Kibana's? 
<img width="767" alt="Screen Shot 2019-08-23 at 14 55 58 PM" src="https://user-images.githubusercontent.com/549577/63616790-56a6c300-c5b6-11e9-8658-7b19a5ff37f2.png">

cchaos commented on src-docs/src/views/elastic_charts/theming_example.js:
```diff
@@ -0,0 +1,122 @@
+            apply them with Charts&apos;{' '}
+            <EuiCode>mergeWithDefaultTheme</EuiCode> function.
+          </p>
+          <p>You&apos;ll find an example of this in the demo below.</p>
```
I'll be able to do this when they fix the `baseTheme` stuff

cchaos replied:
Done.

<img width="976" alt="Screen Shot 2019-08-23 at 17 53 51 PM" src="https://user-images.githubusercontent.com/549577/63625873-fbcd9580-c5ce-11e9-8cf7-64ab5bc5e835.png">

cchaos commented on src-docs/src/views/elastic_charts/sparklines_example.js:
```diff
@@ -0,0 +1,88 @@
+import { EuiSpacer, EuiCode, EuiCodeBlock } from '../../../../src/components';
+
+export const ElasticChartsSparklinesExample = {
+  title: 'Sizing',
```
Yeah it does make more logical sense there, but I wanted to include Sparklines alongside SIzing because I want consumers to know that when they size down super small Sparklines are the way they should be handled. So I didn't want to overload "Creating" with "Sizing".

cchaos commented on src-docs/src/views/elastic_charts/sparklines_example.js:
```diff
@@ -0,0 +1,88 @@
+      text: (
+        <Fragment>
+          <p>
+            As chart containers get smaller, you must re-evaluate your data to
```
It's not when the **user** resizes the chart, it's when the **consumer** places the chart in smaller sized panels/containers. I'll try something along those lines

cchaos commented on src-docs/src/views/elastic_charts/sparklines_example.js:
```diff
@@ -0,0 +1,88 @@
+      demo: <Sizes />,
+    },
+    {
+      title: 'Sparklines',
```
Same reason as above.

cchaos commented on src-docs/src/routes.js:
```diff
@@ -342,6 +352,15 @@ const navigation = [
     ].map(example => createExample(example)),
   },
+  {
+    name: 'Elastic-Charts',
```
I kept Sparklines with sizing but moved it to the second item just below Creating charts

thompsongl commented:
cchaos/eui#24 should resolve the lint errors.

cchaos commented on src/themes/charts/themes.ts:
```diff
@@ -1,13 +1,27 @@
+  PartialTheme,
+  GridLineConfig,
+  LineAnnotationStyle,
+} from '@elastic/charts';
```
@markov00 @thompsongl I've converted this file to TS, would you both mind just giving it a glance?

thompsongl commented on scripts/compile-eui.js:
```diff
@@ -55,7 +55,17 @@ function compileBundle() {
+
+  console.log('Building chart theme module...');
+  execSync(
+    'webpack src/themes/charts/themes.js -o dist/eui_charts_theme.js --output-library-target="commonjs" --config=src/webpack.config.js',
```
```suggestion
    'webpack src/themes/charts/themes.ts -o dist/eui_charts_theme.js --output-library-target="commonjs" --config=src/webpack.config.js',
```

Now that you've converted it to TS

cchaos replied:
done

cchaos commented:
Thanks @markov00, for reference, I've addressed most of your bullet points.

> the copy/paste code will not work 100% on TS because of some missing types. From a JS prospective seems fine except the missing imports of all the components and the formatters
I converted the theme to TS, not too worried about the C/P code at the moment for TS users. Hopefully they'll get their own warnings in order to fix that and the imports.

> each <Chart size={[undefined, 200]}> and similar props can be rewritten in a more “nice” way as size={{height:200}}
Fixed

> it’s better to use enums for the following props: xScaleType, yScaleType, on every Position (left, right, top, bottom), AnnotationDomainType and TooltipType
I only found one instance that wasn't using an enum which was for a `xScaleType` prop. Do you see more?

> we will change in some future version but all the ids needs to be created with getSpecId, getAxisId, getAnnotationId on TS world
I saw in your [Storybook that you're just using straight strings for id's](https://d.pr/free/i/k9pTbG). You may want to then look into your documentation around that.

> on the copied code the data is always written as data={TIME_DATA=[[0,1],[1,2]]} is that intentional?
Fixed

> I think it's missing the histogram chart, that is best suited for all our time series in kibana as we are, most of the time, showing buckets of aggregated values...
I didn't realize there was a particular chart type for that use case. I'll agree it would be good to have a documentation example solely for that. I think it's out of scope right now, but created a github issue to follow up on. #2254

cchaos commented:
Jenkins, test this

thompsongl commented:
Looking back, I don't think a single CI run has succeeded. Even before the lint errors were resolved, CI was failing before it even got to eslint.

> 14:08:17 error @elastic/charts@10.1.1: The engine "node" is incompatible with this module.  Expected version "10.15.2". Got "8.16.1"
> 14:08:17 error Found incompatible module

thompsongl commented:
@chandlerprall Does the CI include devDeps when it installs?

chandlerprall commented:
@cchaos pushed fix for the duplicate react versions, however elastic-charts is entering an infinite loop in IE11, both in these docs and in their published storyboard at elastic.github.com/elastic-charts , looks like the loop is caused by the resize observer /cc @markov00 @nickofthyme

nickofthyme commented:
Thanks @chandlerprall I'll take a look

cchaos commented:
Thanks @chandlerprall !

chandlerprall commented:
@cchaos I've updated the @elastic/elastic-charts dependency to latest version and made a change that I expect to let CI run against this branch.

cchaos commented:
Thank you to all who made this possible 🥇
